File: library/cpp/number/mod_pow.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

/* snippet part start */
const int MOD = 1000000007;

/* x^nを高速に求める */
int mod_pow(int x, int n) {
	int res = 1;
	while (n > 0) {
		if (n & 1) res = res*x%MOD;  // 最下位ビットが立ってる時にx^(2^i)を掛ける
		x = x*x%MOD;              // xを順次二乗していく
		n >>= 1;    // 右ビットシフト
	}
	return res;
}
/* snippet part end */


int main(){
    int n;
    cin >> n;

    for(ll x=2; x<n; x++){
        cout << mod_pow(x,n) <<" "<< x%n << endl;
    }
}

File: library/rust/src/number/mod_pow.rs
// unverified

const MOD:i64 = 1e9 as i64 + 7;

// x^n を高速に求める
fn mod_pow(x: i64, n: i64) -> i64 {
    let mut x = x;
    let mut n = n;
    let mut res = 1;
    while n > 0 {
        if n & 1 > 0 {
            res = res * x % MOD;
        }
        x = x * x % MOD;
        n >>= 1;
    }
    res
}


fn main() {
    let x = 2;
    let n = 10;
    assert_eq!(1024, mod_pow(2,10));
}